daml-lf/engine/src/test/scala/com/digitalasset/daml/lf/engine/EngineTest.scala

```diff
@@ -2434,4 +2562,107 @@ object EngineTest {
     )
   }
 
+  def unusedDisclosedContractsNotSavedToLedger(
+      cmd: speedy.Command,
+      unusedDisclosedContract: DisclosedContract,
+      usedDisclosedContract: DisclosedContract,
+  ): Assertion = {
+    val result = suffixLenientEngine
+      .interpretCommands(
+        validating = false,
+        submitters = Set(alice),
+        readAs = Set.empty,
+        commands = ImmArray(cmd),
+        ledgerTime = Time.Timestamp.now(),
+        submissionTime = Time.Timestamp.now(),
+        seeding = InitialSeeding.TransactionSeed(hash(s"$cmd")),
+        disclosures = ImmArray(unusedDisclosedContract, usedDisclosedContract),
+      )
+      .consume(_ => None, lookupPackage, lookupKey)
+
+    inside(result) { case Right((transaction, metadata)) =>
+      transaction should haveDisclosedInputContracts(usedDisclosedContract)
+      metadata should haveDisclosedContracts(usedDisclosedContract)(preprocessor)
+    }
+  }
+
+  @SuppressWarnings(
+    Array(
+      "org.wartremover.warts.JavaSerializable",
+      "org.wartremover.warts.Product",
+      "org.wartremover.warts.Serializable",
+    )
+  )
+  def haveDisclosedContracts(
+      disclosedContracts: DisclosedContract*
+  )(preprocessor: preprocessing.Preprocessor): Matcher[Tx.Metadata] =
+    Matcher { metadata =>
+      val expectedResult = ImmArray(disclosedContracts: _*)
+      val actualResult = metadata.disclosures
+        .map(_.unversioned)
+        .map(preprocessor.commandPreprocessor.unsafePreprocessDisclosedContract)
+      val debugMessage = Seq(
+        s"expected but missing contract IDs: ${expectedResult.filter(!actualResult.toSeq.contains(_)).map(_.contractId.value)}",
+        s"unexpected but found contract IDs: ${actualResult.filter(!expectedResult.toSeq.contains(_)).map(_.contractId)}",
+      ).mkString("\n  ", "\n  ", "")
+
+      MatchResult(
+        expectedResult == actualResult,
+        s"Failed with unexpected disclosed contracts: $expectedResult != $actualResult $debugMessage",
+        s"Failed with unexpected disclosed contracts: $expectedResult == $actualResult",
+      )
+    }
+
+  def haveDisclosedInputContracts(
+      disclosedContracts: DisclosedContract*
+  ): Matcher[VersionedTransaction] =
+    Matcher { transaction =>
+      val expectedResult = Set(disclosedContracts: _*).map(_.contractId.value)
+      val actualResult = transaction.inputContracts
```

Comment:
This seems a bit of a weird test case. You are testing that the input contracts are _only_ the disclosed contracts. But in general that’s absolutely not true. I can have a transaction that has non-disclosed contracts as well.

What are we actually trying to test for here?

Reply:
> You are testing that the input contracts are only the disclosed contracts. 

Exactly.

As this is intended for a specific set of disclosed contract testing (and not for more general testing), maybe I should flag that by enclosing these matchers in an object?